cmd/peer-rest-client.go

```diff
@@ -190,6 +199,322 @@ func (client *peerRESTClient) NetworkInfo() (info madmin.ServerNetworkHardwareIn
 	return info, err
 }
 
+type networkOverloadedErr struct{}
+
+var networkOverloaded networkOverloadedErr
+
+func (n networkOverloadedErr) Error() string {
+	return "network overloaded"
+}
+
+type progressReader struct {
+	r            io.Reader
+	progressChan chan int64
+}
+
+func (p *progressReader) Read(b []byte) (int, error) {
+	n, err := p.r.Read(b)
+	if err != nil && err != io.EOF {
+		return n, err
+	}
+	p.progressChan <- int64(n)
+	return n, err
+}
+
+func (client *peerRESTClient) doNetOBDTest(ctx context.Context, dataSize int64, threadCount uint) (info madmin.NetOBDInfo, err error) {
+	latencies := []float64{}
+	throughputs := []float64{}
+
+	buf := make([]byte, dataSize)
+
+	buflimiter := make(chan bool, threadCount)
+	errChan := make(chan error, threadCount)
+
+	totalTransferred := int64(0)
```

Comment:
Neither is true.

> I see an intention of trying to determine the average speed while the operation was going, but the totalTransferred is an ahead-of-time counter 

`totalTransferred` is not an ahead-of-time counter. `totalTransferred` is only incremented after the buffer has been emptied of `n` bytes.

>   that does not say much about when the transfer is taking place. 

Recording the start and stop time alone is actually the wrong approach. I'll explain why - If you measured start and stop time of `n` operations, you'd end up with n posibbly-overlapping intervals.

```
1-3
2-4
2-6
3-5
...
```

You'll then have to break them apart into non-overlapping intervals and compute the number of active threads in each of those.

```
Interval      |   threads
1-2           |    1
2-3           |    2
3-4           |    3
4-5           |    2
5-6           |    1 
``` 

Using this data, the throughput can only be computed with the assumption that data is being transferred at a constant rate of `partial_size/sec` for a given thread in a given interval. 
where partial_size is computed as 

```
    partial_size = obj_size * number of seconds to complete entire op / number of seconds in interval
```

The throughput of the entire interval can only be computed as the sum of throughputs of each thread in that interval. The issue with this approach is that it assumes the throughput for a given object in a given thread is uniformly distributed. This actually leads to skewness. 

In contrast, my approach measures throughput in terms of total bytes transferred between two points in time. There is simply no need to count the number of active threads or the non-overlapping intervals from each of the individual overlapping operations. This approach, In addition to being simple and elegant, it always leads to accurate results. 

> The actual readahead of the transfer buffer will apply a significant skew to the numbers AFAICT.

There is no readahead. 

> Furthermore this also relies on the transferChan being emptied at once - this could at least be remedied by having the counter inside the progressReader and update it at once when the read occurs.

I don't think you fully understand the approach. Please refer to my explanation above. 

> Furthermore, once there is less than threadCount transfers running you are basically testing something new and these should be discarded.

If I relied on the `threadCount` being emptied at once, the results would be skewed every single time. A simple sanity check will prove otherwise. In my throughput computation, I'm not multiplying the size of an object with the number of active threads. I believe your assumption is coming from the approach in warp and it is bound to skewness. Please refer to my explanation above.

Reply:
> `totalTransferred` is not an ahead-of-time counter. `totalTransferred` is only incremented after the buffer has been emptied of `n` bytes.

Correct, didn't read it correctly.

However, get rid of the lag imposed by the `transferChan` channel. Right now your `before`/`after` relies on everything in `transferChan` to have been applied. But you can just get rid of it.

> In contrast, my approach measures throughput in terms of total bytes transferred between two points in time.

> There is simply no need to count the number of active thread.

My main objection is that this assumes that 1 connection can saturate the NIC. We know that isn't true which is why you have several goroutines running. So we should consider the number of requests, or at least reject the ones at the end where less are running.

Reply:
Again, neither is true. Maybe we should get on a call, so I can explain this to you better.